Full C-Set implementation

mehalter opened the pull request:
Working on implementing several types of Petri nets (with/without labels and/or rates/initial conditions)

olynch commented:
This looks really cool! Let's talk about this at the Thursday meeting.

olynch commented:
Oh, also it's interesting that you have different data for rates and concentration. Is this because they are conceptually in different units?

mehalter commented:
I separated those out because there are cases where they are stored in different types. For example, concentration is limited to natural numbers while rates can be real numbers

olynch commented:
Ah! In a stochastic framework, I see!

mehalter commented:
This latest commit moves the open petri implementation to using the CSet data structures. Lastly, I just need to update the example files to use this updated API

mehalter commented:
Updated the examples and added nearly exhaustive testing. Going to start working on fully testing the vectorfield generation of the CSet datastructure and add full support for DifferentialEquations.jl natively. Once that is done I think this will be ready to merge :partying_face:

epatters commented on src/AlgebraicPetri.jl:
```diff
@@ -121,19 +109,18 @@ end
         # reimplementation of pushout of Span{FinSetFunc, FinSetFun}
```
This shouldn't be necessary anymore since calling `pushout` on FinFunctions now returns a [`CompositePushout`](https://github.com/AlgebraicJulia/Catlab.jl/blob/414b9787f90014c33a7b125d50554ebc044d4513/src/categorical_algebra/Limits.jl#L280). You can fish the coequalizer out of it if you have to.

epatters replied:
In principle, we shouldn't have to do *that* either because it breaks the abstraction boundary for the pushout. See the docs for [`CompositePullback`](https://github.com/AlgebraicJulia/Catlab.jl/blob/414b9787f90014c33a7b125d50554ebc044d4513/src/categorical_algebra/Limits.jl#L249). I was going to say that you should just use the universal property of the pushout, which is now implemented, but reminding myself about decorated cospans I see that it actually uses properties of the both coproduct *and* pushout. So I see two possibilities here:

1. Keep what you currently have because you actually need both the coproduct and coequalizer/pushout.
2. Compute the coproduct *and* the pushout; take the legs of the pushout, which are morphisms into the pushout object; then apply the universal property of the coproduct to these morphisms, giving a morphism from the coproduct object to the pushout object.

epatters replied:
Anyway, I figure this will get resolved anyway when we have a generic implementation of decorated/structured cospans for C-sets, so I wouldn't worry about it too much.

mehalter replied:
Yeah, I think it would be ok to leave it as is until the generic structured cospans is implemented for 𝒞-sets

mehalter commented:
Just finished implementing `vectorfield` generation for both unattributed and attributed petri 𝒞-sets with support for variable transition rates that extend petri nets to arbitrary dynamical systems! I think this is ready to be reviewed and merged!

epatters commented on src/Types.jl:
```diff
@@ -0,0 +1,258 @@
+using LinearAlgebra: mul!
+import Petri: Model
+
+vectorify(n) = begin
```
Is this function needed? The C-sets API is vectorized, so you should be able to mix scalars and vectors without worrying about it.

mehalter replied:
Ah this helper function is for the constructors to help me iterate over both arrays and scalars. I was running into some weird behavior where I was getting a 0-dimensional array in one of the constructors. It's not to help with the C-Set API

epatters commented on src/Types.jl:
```diff
@@ -0,0 +1,258 @@
+add_output!(p::AbstractPetriNet,t,s;kw...) = add_part!(p,:O;ot=t,os=s,kw...)
+add_outputs!(p::AbstractPetriNet,n,t,s;kw...) = add_parts!(p,:O,n;ot=t,os=s,kw...)
+
+# Note: although indexing makes this pretty fast, it is often faster to bulk-convert
```
Out of curiosity, did you run benchmarks? At some point we should benchmark C-sets to make sure they perform as well in reality as they should in theory.

mehalter replied:
I haven't yet, with it fully implemented I am interested to see how it performs compared to the Petri.jl implementation.

epatters commented on src/Types.jl:
```diff
@@ -0,0 +1,258 @@
+inputs(p::AbstractPetriNet,t) = subpart(p,incident(p,t,:it),:is)
+outputs(p::AbstractPetriNet,t) = subpart(p,incident(p,t,:ot),:os)
+
+struct TransitionMatrices
```
We should add a generic function to Catlab that gives you the (multi)relation matrix given by two parts of a C-set (or higher-order tensor given by more parts).

mehalter replied:
Oh yeah that's a great idea. I can start working on that some as well

epatters reviewed (commented):
Looks great!
